materialize-kafka/src/configuration.rs

```diff
@@ -294,6 +338,85 @@ impl JsonSchema for Resource {
     }
 }
 
+pub struct AckTrackingContext {
+    pending_messages: AtomicUsize,
+    completed_messages: AtomicUsize,
+    has_failure: AtomicBool,
+    state: Mutex<Option<KafkaError>>,
+    notify: Condvar,
+}
+
+impl AckTrackingContext {
+    pub fn new() -> Self {
+        Self {
+            pending_messages: AtomicUsize::new(0),
+            completed_messages: AtomicUsize::new(0),
+            has_failure: AtomicBool::new(false),
```

Comment:
Using this `has_failure` seems like overkill to me. The `wait_for_all_successful_acks` loop should only get one notification as I am reading the code, so it seems like it would be enough to just do the `if let Some(ref error) = *guard` check at the top of the loop? That would also alleviate a secondary question I have, which is what happens if `has_failure` is true but `*guard` is `None`. Of course this is not a possible condition but it seems more complicated than it needs to be.

Reply:
Yep, this is left over from an earlier factoring where we'd wake the thread every message. Removed.